Idrall/Value.idr

```diff
@@ -4,157 +4,156 @@ import Idrall.Expr
 import Idrall.Error
 
 mutual
-  public export
-  data Normal = Normal' Ty Value
-
   public export
   Ty : Type
   Ty = Value
 
   -- Values
   public export
   data Value
-    = VLambda Ty Closure
+    = VConst U
+    | VVar Name Int
+    | VPrimVar
+    | VApp Value Value
+    | VLambda Ty Closure
     | VHLam HLamInfo (Value -> Either Error Value)
     | VPi Ty Closure
-    | VHPi String Value (Value -> Value)
-    | VEquivalent Value Value
-    | VAssert Value
-    | VConst U
+    | VHPi Name Value (Value -> Either Error Value)
+
     | VBool
     | VBoolLit Bool
-    | VInteger
-    | VIntegerLit Integer
+    | VBoolAnd Value Value
+
     | VNatural
     | VNaturalLit Nat
+    | VNaturalIsZero Value
+
+    | VInteger
+    | VIntegerLit Integer
+    | VIntegerNegate Value
+
     | VDouble
     | VDoubleLit Double
-    | VList Ty
-    | VListLit (Maybe Ty) (List Value)
+
     | VText
     | VTextLit VChunks
+
+    | VList Ty
+    | VListLit (Maybe Ty) (List Value)
+    | VListAppend Value Value
+    | VListHead Value Value
+
     | VOptional Ty
     | VNone Ty
     | VSome Ty
+
+    | VEquivalent Value Value
+    | VAssert Value
+
     | VRecord (SortedMap FieldName Value)
     | VRecordLit (SortedMap FieldName Value)
     | VUnion (SortedMap FieldName (Maybe Value))
+    | VCombine Value Value
+    | VCombineTypes Value Value
     | VInject (SortedMap FieldName (Maybe Value)) FieldName (Maybe Value) -- TODO proof that key is in SM?
-    | VPrimVar
-    | VNeutral Ty Neutral
 
   public export
-  Env : Type -- Now a type alias
-  Env = List (Name,Value)
-  %name Env env, env1, env2
+  data Env
```

Comment:
@Danten just realised, this representation for Env may fix the issues we ran into with the scope checker. When weakening closures, we can now weaken their Env without having to create a value. Something to think about, for whenever the next attempt is 🤔